0.37.0: InvalidPathException

When I invoke `java -jar C:/ktlint/ktlint-0.37.0-all.jar src/**/*.kt` I get the following stacktrace:

```
Exception in thread "main" java.util.concurrent.ExecutionException: java.nio.file.InvalidPathException: Illegal char <<> at index 0: <stdin>
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at com.pinterest.ktlint.KtlintCommandLine.parallel(Main.kt:532)
        at com.pinterest.ktlint.KtlintCommandLine.parallel$default(Main.kt:506)
        at com.pinterest.ktlint.KtlintCommandLine.lintFiles(Main.kt:265)
        at com.pinterest.ktlint.KtlintCommandLine.run(Main.kt:233)
        at com.pinterest.ktlint.Main.main(Main.kt:62)
Caused by: java.nio.file.InvalidPathException: Illegal char <<> at index 0: <stdin>
        at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:229)
        at java.base/java.nio.file.Path.endsWith(Path.java:435)
        at com.pinterest.ktlint.core.internal.EditorConfigLoader.loadPropertiesForFile(EditorConfigLoader.kt:47)
        at com.pinterest.ktlint.core.KtLint.lint(KtLint.kt:143)
        at com.pinterest.ktlint.internal.FileUtilsKt.lintFile(FileUtils.kt:51)
        at com.pinterest.ktlint.KtlintCommandLine.process(Main.kt:358)
        at com.pinterest.ktlint.KtlintCommandLine.access$process(Main.kt:116)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$2$1.call(Main.kt:257)
        at com.pinterest.ktlint.KtlintCommandLine$lintFiles$2$1.call(Main.kt:116)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
```

Environment:
* Version of ktlint used: 0.37.0
* Name and version: ktlint-0.37.0-all.jar
* Operating System and version: Windows 10 Enterprise

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,11 @@
 All notable changes to this project will be documented in this file.  
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## [Unreleased]
+
+### Fixed
+- Invalid path exception error on windows machines when loading properties from .editorconfig ([#761](https://github.com/pinterest/ktlint/issues/761))
+
 ## [0.37.0] - 2020-06-02
 
 Thank you to [Tapchicoma](https://github.com/Tapchicoma) and [romtsn](https://github.com/romtsn) for all their hard work on this release!

--- a/ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/KtLint.kt
+++ b/ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/KtLint.kt
@@ -2,6 +2,8 @@ package com.pinterest.ktlint.core
 
 import com.pinterest.ktlint.core.ast.prevLeaf
 import com.pinterest.ktlint.core.internal.EditorConfigLoader
+import java.nio.file.FileSystems
+import java.nio.file.Path
 import java.nio.file.Paths
 import java.util.ArrayList
 import java.util.HashSet
@@ -50,7 +52,7 @@ object KtLint {
     private val psiFileFactory: PsiFileFactory
     private val nullSuppression = { _: Int, _: String, _: Boolean -> false }
 
-    private val editorConfigLoader = EditorConfigLoader()
+    private val editorConfigLoader = EditorConfigLoader(FileSystems.getDefault())
 
     /**
      * @param fileName path of file to lint/format
@@ -71,7 +73,15 @@ object KtLint {
         val script: Boolean = false,
         val editorConfigPath: String? = null,
         val debug: Boolean = false
-    )
+    ) {
+        internal val normalizedFilePath: Path? get() = if (fileName == STDIN_FILE || fileName == null) {
+            null
+        } else {
+            Paths.get(fileName)
+        }
+
+        internal val isStdIn: Boolean get() = fileName == STDIN_FILE
+    }
 
     init {
         // do not print anything to the stderr when lexer is unable to match input
@@ -141,7 +151,8 @@ object KtLint {
         val rootNode = psiFile.node
         // Passed-in userData overrides .editorconfig
         val mergedUserData = editorConfigLoader.loadPropertiesForFile(
-            params.fileName?.let { Paths.get(it) },
+            params.normalizedFilePath,
+            params.isStdIn,
             params.editorConfigPath?.let { Paths.get(it) },
             params.debug
         ) + params.userData
@@ -322,7 +333,8 @@ object KtLint {
         val rootNode = psiFile.node
         // Passed-in userData overrides .editorconfig
         val mergedUserData = editorConfigLoader.loadPropertiesForFile(
-            params.fileName?.let { Paths.get(it) },
+            params.normalizedFilePath,
+            params.isStdIn,
             params.editorConfigPath?.let { Paths.get(it) },
             params.debug
         ) + params.userData

--- a/ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/internal/EditorConfigLoader.kt
+++ b/ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/internal/EditorConfigLoader.kt
@@ -1,9 +1,8 @@
 package com.pinterest.ktlint.core.internal
 
-import com.pinterest.ktlint.core.KtLint.STDIN_FILE
 import java.nio.charset.StandardCharsets
+import java.nio.file.FileSystem
 import java.nio.file.Path
-import java.nio.file.Paths
 import org.ec4j.core.EditorConfigLoader
 import org.ec4j.core.Resource
 import org.ec4j.core.ResourcePropertiesService
@@ -13,7 +12,9 @@ import org.ec4j.core.ResourcePropertiesService
  *
  * Contains internal in-memory cache to speedup lookup.
  */
-class EditorConfigLoader {
+class EditorConfigLoader(
+    private val fs: FileSystem
+) {
     private val cache = ThreadSafeEditorConfigCache()
     private val editorConfigLoader = EditorConfigLoader.default_()
     private val propService = ResourcePropertiesService.builder()
@@ -25,7 +26,9 @@ class EditorConfigLoader {
      * Loads applicable properties from `.editorconfig`s for given file.
      *
      * @param filePath path to file that would be checked.
-     * Could end with [STDIN_FILE] to indicate that _stdin_ would be checked
+     * @param isStdIn indicates that checked content comes from input.
+     * Setting this to `true` overrides [filePath] and uses `.kt` pattern to load properties
+     * from current folder `.editorconfig` files.
      * @param alternativeEditorConfig alternative to current [filePath] location where `.editorconfig` files should be
      * looked up
      * @param debug pass `true` to enable some additional debug output
@@ -36,23 +39,26 @@ class EditorConfigLoader {
      */
     fun loadPropertiesForFile(
         filePath: Path?,
+        isStdIn: Boolean = false,
         alternativeEditorConfig: Path? = null,
         debug: Boolean = false
     ): Map<String, String> {
-        if (filePath == null || SUPPORTED_FILES.none { filePath.toString().endsWith(it) }) {
+        if (!isStdIn &&
+            (filePath == null || SUPPORTED_FILES.none { filePath.toString().endsWith(it) })
+        ) {
             return emptyMap()
         }
 
         val normalizedFilePath = when {
-            filePath.endsWith(STDIN_FILE) ->
-                Paths
-                    .get(".")
+            isStdIn ->
+                fs
+                    .getPath(".")
                     .toAbsolutePath()
                     .resolve("stdin${SUPPORTED_FILES.first()}")
             alternativeEditorConfig != null ->
                 alternativeEditorConfig
                     .toAbsolutePath()
-                    .resolve("${filePath.last()}")
+                    .resolve("${filePath!!.last()}")
             else -> filePath
         }
 
@@ -65,7 +71,7 @@ class EditorConfigLoader {
             .properties
             .mapValues { it.value.sourceValue }
             .run {
-                if (!filePath.endsWith(STDIN_FILE)) {
+                if (!isStdIn) {
                     plus(FILE_PATH_PROPERTY to filePath.toString())
                 } else {
                     this
@@ -87,8 +93,7 @@ class EditorConfigLoader {
          */
         internal val SUPPORTED_FILES = arrayOf(
             ".kt",
-            ".kts",
-            STDIN_FILE
+            ".kts"
         )
     }
 }

--- a/ktlint-core/src/test/kotlin/com/pinterest/ktlint/core/internal/EditorConfigLoaderTest.kt
+++ b/ktlint-core/src/test/kotlin/com/pinterest/ktlint/core/internal/EditorConfigLoaderTest.kt
@@ -2,7 +2,6 @@ package com.pinterest.ktlint.core.internal
 
 import com.google.common.jimfs.Configuration
 import com.google.common.jimfs.Jimfs
-import com.pinterest.ktlint.core.KtLint.STDIN_FILE
 import java.nio.file.FileSystem
 import java.nio.file.Files
 import java.nio.file.Path
@@ -13,7 +12,7 @@ import org.junit.Test
 
 internal class EditorConfigLoaderTest {
     private val tempFileSystem = Jimfs.newFileSystem(Configuration.forCurrentPlatform())
-    private val editorConfigLoader = EditorConfigLoader()
+    private val editorConfigLoader = EditorConfigLoader(tempFileSystem)
 
     private fun FileSystem.normalizedPath(path: String): Path {
         val root = rootDirectories.joinToString(separator = "/")
@@ -221,21 +220,24 @@ internal class EditorConfigLoaderTest {
 
     @Test
     fun `Should return properties for stdin from current directory`() {
-        val projectDir = "/project"
         @Language("EditorConfig") val editorconfigFile =
             """
             [*.{kt,kts}]
             insert_final_newline = true
             disabled_rules = import-ordering
             """.trimIndent()
-        tempFileSystem.writeEditorConfigFile(projectDir, editorconfigFile)
+        tempFileSystem.writeEditorConfigFile(".", editorconfigFile)
 
-        val lintFile = tempFileSystem.normalizedPath(projectDir).resolve(STDIN_FILE)
-        val parsedEditorConfig = editorConfigLoader.loadPropertiesForFile(lintFile)
+        val parsedEditorConfig = editorConfigLoader.loadPropertiesForFile(null, isStdIn = true, debug = true)
 
         assertThat(parsedEditorConfig).isNotEmpty
         assertThat(parsedEditorConfig).doesNotContainKey(EditorConfigLoader.FILE_PATH_PROPERTY)
-        // It is not possible to reliably check all loaded properties for stdin as it will depend on the machine
+        assertThat(parsedEditorConfig).isEqualTo(
+            mapOf(
+                "insert_final_newline" to "true",
+                "disabled_rules" to "import-ordering"
+            )
+        )
     }
 
     @Test

--- a/ktlint/src/main/kotlin/com/pinterest/ktlint/internal/IntellijIDEAIntegration.kt
+++ b/ktlint/src/main/kotlin/com/pinterest/ktlint/internal/IntellijIDEAIntegration.kt
@@ -6,6 +6,7 @@ import java.io.ByteArrayInputStream
 import java.io.ByteArrayOutputStream
 import java.io.IOException
 import java.nio.charset.Charset
+import java.nio.file.FileSystems
 import java.nio.file.Files
 import java.nio.file.NoSuchFileException
 import java.nio.file.Path
@@ -27,7 +28,8 @@ object IntellijIDEAIntegration {
         if (!Files.isDirectory(workDir.resolve(".idea"))) {
             throw ProjectNotFoundException()
         }
-        val editorConfig: Map<String, String> = EditorConfigLoader().loadPropertiesForFile(Paths.get("."))
+        val editorConfig: Map<String, String> = EditorConfigLoader(FileSystems.getDefault())
+            .loadPropertiesForFile(null, isStdIn = true)
         val indentSize = editorConfig["indent_size"]?.toIntOrNull() ?: 4
         val continuationIndentSize = editorConfig["continuation_indent_size"]?.toIntOrNull() ?: 4
         val updates = if (local) {

